substrate/frame/transaction-payment/skip-feeless-payment/src/lib.rs

```diff
@@ -0,0 +1,145 @@
+//!
+//!
+//! ## Integration
+//!
+//! This pallet wraps an existing transaction payment pallet. This means you should both pallets
+//! in your `construct_runtime` macro and include this pallet's
+//! [`SignedExtension`] ([`SkipCheckIfFeeless`]) that would accept the existing one as an argument.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Decode, Encode};
+use frame_support::{
+	dispatch::{CheckIfFeeless, DispatchResult},
+	traits::{IsType, OriginTrait},
+};
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{DispatchInfoOf, PostDispatchInfoOf, SignedExtension},
+	transaction_validity::TransactionValidityError,
+};
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A transaction fee was skipped.
+		FeeSkipped { who: T::AccountId },
+	}
+}
+
+/// A [`SignedExtension`] that skips the wrapped extension if the dispatchable is feeless.
+#[derive(Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct SkipCheckIfFeeless<T: Config, S: SignedExtension>(pub S, sp_std::marker::PhantomData<T>);
+
+impl<T: Config, S: SignedExtension> sp_std::fmt::Debug for SkipCheckIfFeeless<T, S> {
+	#[cfg(feature = "std")]
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		write!(f, "SkipCheckIfFeeless<{:?}>", self.0.encode())
+	}
+	#[cfg(not(feature = "std"))]
+	fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		Ok(())
+	}
+}
+
+impl<T: Config + Send + Sync, S: SignedExtension> SkipCheckIfFeeless<T, S> {
+	/// utility constructor. Used only in client/factory code.
+	pub fn from(s: S) -> Self {
+		Self(s, sp_std::marker::PhantomData)
+	}
+}
+
+impl<T: Config + Send + Sync, S: SignedExtension<AccountId = T::AccountId>> SignedExtension
+	for SkipCheckIfFeeless<T, S>
+where
+	S::Call: CheckIfFeeless<Origin = frame_system::pallet_prelude::OriginFor<T>>,
+{
+	type AccountId = T::AccountId;
+	type Call = S::Call;
+	type AdditionalSigned = S::AdditionalSigned;
+	type Pre = (Self::AccountId, Option<<S as SignedExtension>::Pre>);
+	const IDENTIFIER: &'static str = "SkipCheckIfFeeless";
+
+	fn additional_signed(&self) -> Result<Self::AdditionalSigned, TransactionValidityError> {
+		self.0.additional_signed()
+	}
+
+	fn pre_dispatch(
+		self,
+		who: &Self::AccountId,
+		call: &Self::Call,
+		info: &DispatchInfoOf<Self::Call>,
+		len: usize,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		if call.is_feeless(&<T as frame_system::Config>::RuntimeOrigin::signed(who.clone())) {
+			Ok((who.clone(), None))
```

Comment:
I think it is worth taking a moment for yourself to conclude the following, and document it for external users as well: 

Even if a transaction is feeless, it still consumes weight and they two are totally independent.